File: jscomp/runtime/caml_exceptions.ts
// Js_of_ocaml runtime support
// http://www.ocsigen.org/js_of_ocaml/
// Copyright (C) 2014 Jérôme Vouillon, Hugo Heuzard, Andy Ray
// Laboratoire PPS - CNRS Université Paris Diderot
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, with linking exception;
// either version 2.1 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

//  Copyright (c) 2015 Bloomberg LP. All rights reserved. 
// Hongbo Zhang (hzhang295@bloomberg.net)              

"use strict";
/**
 * Exported for better inlining
 * It's common that we have <code> a = caml_set_oo_id([248,"string",0])</code>
 * This can be inlined as <code> a = caml_set_oo_id([248,"tag", caml_oo_last_id++])</code>
 * @type {number}
 */
var caml_oo_last_id = 0;


function caml_set_oo_id (b) {
    b[2]=caml_oo_last_id++;
    return b;
}

var Out_of_memory = [248, "Out_of_memory", 0];
var Sys_error = [248, "Sys_error", -1];
var Failure = [248, "Failure", -2];
var Invalid_argument = [248, "Invalid_argument", -3];
var End_of_file = [248, "End_of_file",-4];
var Division_by_zero = [248, "Division_by_zero", -5];
var Not_found = [248, "Not_found", -6];
var Match_failure = [248, "Match_failure", -7];
var Stack_overflow = [248, "Stack_overflow",-8];
var Sys_blocked_io = [248, "Sys_blocked_io", -9];
var Assert_failure = [248, "Assert_failure", -10];
var Undefined_recursive_module = [248, "Undefined_recursive_module", -11];

function caml_raise_constant (tag) { throw tag; }


function caml_return_exn_constant (tag) { return tag; }


function caml_raise_with_arg (tag, arg) { throw [0, tag, arg]; }

function caml_raise_with_string (tag, msg) {
    caml_raise_with_arg (tag, msg);
}

function caml_raise_sys_error (msg) {
    caml_raise_with_string(Sys_error, msg);
}

function caml_invalid_argument (msg) {
    caml_raise_with_string(Invalid_argument, msg);
}

function caml_raise_end_of_file () {
    caml_raise_constant( End_of_file);
}

function caml_raise_zero_divide () {
    caml_raise_constant(Division_by_zero);
}

function caml_raise_not_found () {
    caml_raise_constant(Not_found);
}



function caml_array_bound_error () {
    caml_invalid_argument("index out of bounds");
}


/**
 * This function does not exist in runtime
 * However, we still need here since we need write some low-level code which
 * needs such utilities, if we write such low level code in OCaml itself, maybe
 * we can remove it
 * void caml_failwith(char const *msg)
 * asmrun/fail.c
 * @param msg
 */
function caml_failwith (msg : string) {
    caml_raise_with_string(Failure, msg);
}


//Provides: caml_wrap_exception const (const)
//Requires: caml_global_data,caml_js_to_string,caml_named_value
//Requires: caml_return_exn_constant
//function caml_wrap_exception(e) {
//  if(e instanceof Array) return e;
//  //Stack_overflow: chrome, safari
//  if(RangeError !== undefined
//     && e instanceof RangeError
//     && e.message
//     && e.message.match(/maximum call stack/i))
//    return caml_return_exn_constant(caml_global_data["Stack_overflow"]);
//  //Stack_overflow: firefox
//  if(InternalError !== undefined
//     && e instanceof InternalError
//     && e.message
//     && e.message.match(/too much recursion/i))
//    return caml_return_exn_constant(caml_global_data["Stack_overflow"]);
//  //Wrap Error in Js.Error exception
//  if(e instanceof Error)
//    return [0,caml_named_value("jsError"),e];
//  //fallback: wrapped in Failure
//  return [0,caml_global_data["Failure"],String(e)];
//}


export {
    Out_of_memory, Sys_error, Failure,
    Invalid_argument, End_of_file, Division_by_zero,
    Not_found, Match_failure, Stack_overflow, Sys_blocked_io,
    Assert_failure, Undefined_recursive_module,

    caml_oo_last_id,
    caml_set_oo_id,
    caml_raise_sys_error,
    caml_failwith, // Internal
    caml_invalid_argument, // Internal
    caml_array_bound_error ,// Internal
    caml_raise_zero_divide,
    caml_raise_not_found,
    caml_raise_with_arg
}

File: jscomp/runtime/caml_exceptions.ml
(* Copyright (C) 2015-2016 Bloomberg Finance L.P.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

type t = {
  id : string [@bs.as "ExceptionID"];

} 


(** 
   Could be exported for better inlining
   It's common that we have 
   {[ a = caml_set_oo_id([248,"string",0]) ]}
   This can be inlined as 
   {[ a = caml_set_oo_id([248,"tag", caml_oo_last_id++]) ]}
*)

let%private id = ref 0


(* see  #251
   {[
     CAMLprim value caml_set_oo_id (value obj) {
       Field(obj, 1) = oo_last_id;
       oo_last_id += 2;
       return obj;
     }

   ]}*)
let caml_set_oo_id (b : Caml_obj_extern.t)  : Caml_obj_extern.t = 
  Caml_obj_extern.set_field (Caml_obj_extern.repr b) 1 (Caml_obj_extern.repr id.contents);
  id .contents <-  id.contents  + 1; 
  b
(* FXIME: this is only relevant to OO module now *)


let create (str : string) : string = 
  id .contents <-  id.contents + 1;  
  str ^ "/" ^(Obj.magic (id.contents : int) : string)




(**
   This function should never throw
   It could be either customized exception or built in exception 
   Note due to that in OCaml extensible variants have the same 
   runtime representation as exception, so we can not 
   really tell the difference. 

   However, if we make a false alarm, classified extensible variant 
   as exception, it will be OKAY for nested pattern match

   {[
     match toExn x : exn option with 
     | Some _ 
       -> Js.log "Could be an OCaml exception or an open variant"
     (* If it is an Open variant, it will never pattern match, 
        This is Okay, since exception could never have exhaustive pattern match

     *)
     | None -> Js.log "Not an OCaml exception for sure"
   ]}

   However, there is still something wrong, since if user write such code
   {[
     match toExn x with 
     | Some _ -> (* assert it is indeed an exception *)
       (* This assertion is wrong, since it could be an open variant *)
     | None -> (* assert it is not an exception *)
   ]}

   This is not a problem in `try .. with` since the logic above is not expressible, see more design in [destruct_exn.md]
*)
let caml_is_extension (type a ) (e : a) :  bool  =
  if Js.testAny e then false 
  else Js.typeof (Obj.magic e : t) .id = "string"




(**FIXME: remove the trailing `/` *)
let caml_exn_slot_name (x : t) : string = x.id
